### ISSUE TITLE
Imbalance classes

### ISSUE THREAD

**Issue by 32nguyen** [jocicmarko/ultrasound-nerve-segmentation#33]
I am trying to segment imbalanced 3 classes in my brain volume (following this paper: [https://arxiv.org/pdf/1702.05970.pdf](https://arxiv.org/pdf/1702.05970.pdf) page 7 formula (1) by adding the weight_class. 

My label volume (after to_categorical is (batch_size, img_size^2, 3)). But I think keras's class_weight in fit function does not support 3D tensor.
I think I have to adjust the loss function by creating the weight_volume corresponding to to label volume. My question is if **_I use the dice function, how can I implement this?_**

I would like to post my failed code when using the cross entropy loss function. It works with simple y_pred and y_true that I defined by myself. But it does not work when it embedded to Unet due to placeholder, nontype at first dimension in Unet. I believe that the code has a lot of trouble, but hopefully it will show I idea.  

```
val = np.array([[[1, 0, 0], [0, 1, 0], [0, 1, 0]], 
                   [[0, 0, 1], [0, 0, 1], [1, 0, 0]], 
                   [[1, 0, 0], [1, 0, 0], [0, 0, 1]]])

val1 = np.array([[[1, 0, 0], [0, 1, 0], [0, 1, 0]], 
                   [[0, 0, 1], [0, 0, 1], [1, 0, 0]], 
                   [[1, 0, 0], [1, 0, 0], [0, 0, 1]]])

kvar = K.variable(value=val, dtype='float32', name='example_var')
kvar1 = K.variable(value=val1, dtype='float32', name='example_var')

def cross_entropy_function(y_true, y_pred):
    # y_true and y_pred has shape (batch_size, img_size*_img_size, 3)
    shape = K.int_shape(y_true)
    weight0 = K.get_value((shape[0]*shape[1] - K.sum(K.flatten(y_true[:,:,0])))
                          /K.sum(K.flatten(y_true[:,:,0]))-1.0)
    weight1 = K.get_value((shape[0]*shape[1] - K.sum(K.flatten(y_true[:,:,1])))
                          /K.sum(K.flatten(y_true[:,:,1]))-1.0)
    weight2 = K.get_value((shape[0]*shape[1] - K.sum(K.flatten(y_true[:,:,2])))
                          /K.sum(K.flatten(y_true[:,:,2]))-1.0)
    weight_y_true = K.get_value(y_true)
    #print weight0, weight1, weight2
    weight_y_true_0 = np.ones(shape = (shape[0],shape[1],1))
    weight_y_true_1 = np.ones(shape = (shape[0],shape[1],1))
    weight_y_true_2 = np.ones(shape = (shape[0],shape[1],1))
    weight_y_true_0[np.where(weight_y_true[:,:,0]==1)] = weight0 
    weight_y_true_1[np.where(weight_y_true[:,:,1]==1)] = weight1 
    weight_y_true_2[np.where(weight_y_true[:,:,2]==1)] = weight2
    A = np.concatenate((weight_y_true_0, weight_y_true_1), 2)
    A = np.concatenate((A, weight_y_true_2), 2)
    class_weight = K.variable(value=A, dtype='float32')

    # calculate weighted cross entropy
    epsilon = tf.constant(value=1e-4)
    y_true_f = K.flatten(y_true)
    y_pred_f = K.flatten(y_pred) 
    class_weight_f = K.flatten(class_weight)
    L = tf.reduce_sum(class_weight_f*
                      (y_true_f*K.log(y_pred_f+ epsilon) + 
                       (1-y_true_f)*K.log(1-y_pred_f+ epsilon)))
    L = tf.reduce_mean(L)
    #print K.get_value(L)
    return L 

def loss(y_true, y_pred):    
    return -cross_entropy_function(y_true, y_pred)
print K.get_value(loss(kvar, kvar1))

```
